thunder/core/transform_common.py

```diff
@@ -363,3 +368,105 @@ class PostOptimizationTransform(Transform, ABC):
     @abstractmethod
     def transform_trace(self, computation_trace: Trace, **kwargs):
         pass
+
+
+def functionalize_inplace_ops(computation_trace: Trace) -> list[Trace]:
+    """Functionalize in-place ops in ``computation_trace``.
+
+    In thunder, an in-place is an out-of-place or functional op followed by :func:`~thunder.core.prims.copy_`.
+    This function replaces such in-place ops with out-of-place ops.
+    Note that functionalization is not applied, if any of an in-place op's arguments are
+    ``computation_trace.args`` or ``computation_trace.kwargs``.
+
+    For example, :func:`thunder.torch.add_` is represented as a :class:`thunder.core.symbol.BoundSymbol`
+    whose `subsymbols` are :func:`thunder.torch.add` and :func:`thunder.core.prims.copy_`. This function
+    replaces it with a :class:`~thunder.core.symbol.BoundSymbol` of :func:`~thunder.torch.add`.
+    """
+    import thunder.torch
+
+    def is_functionalizable(bsym: BoundSymbol) -> bool:
+        """Has `OpTags.IN_PLACE` and its args are NOT ``computation_trace.args`` nor ``computation_trace.kwargs``."""
+        return (
+            bsym.sym.tags
+            and prims.OpTags.IN_PLACE in bsym.sym.tags
+            and bsym.subsymbols
+            and bsym.subsymbols[-1].sym.id == prims.PrimIDs.COPY_
```

Comment:
we certainly should drop the subsymbols check. This is irrelevant from how this PR is handling functionalization.

Reply:
> This is irrelevant from how this PR is handling functionalization.

why is it irrelevant? Currently in-place bsyms have out-of-place and copy as their subsymbols so I think it fair to check the last sub bound symbol is copy.

Reply:
how can we tell an appropriate output tensor proxy if a bsym doesn't have a copy_ as its last sub bsym, while avoiding having a lot of new tensor proxy names in a trace?

Reply:
oops. sorry I read your implementation wrong earlier... I thought we are doing a blind `torch.xxx_` to `torch.xxx` replacement but that's not the case. You actually are only looking at the last subsymbol and replacing that one entry only.

Reply:
That feels a bit restricted... But a first step is still better then nothing and I'll stop nitpicking on that.

Reply:
> That feels a bit restricted... But a first step is still better then nothing and I'll stop nitpicking on that.

how would it be a bit restricted compared to a blind `torch.foo_` to `torch.foo` replacement?

Reply:
no that's not what I meant.

By `That feels a bit restricted`, I'm referring to the alternative that we functionalize directly at the subsymbol `prim.copy_` level. But again we don't have to do that in this PR.